File: src/main/java/com/skd/sublimacion_api/service/DisenoService.java
package com.skd.sublimacion_api.service;

import com.skd.sublimacion_api.dto.diseno.DisenoPublicoResponse;
import com.skd.sublimacion_api.dto.diseno.DisenoRequest;
import com.skd.sublimacion_api.dto.diseno.DisenoResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface DisenoService {
    DisenoResponse generar(DisenoRequest request, Long usuarioId);
    DisenoResponse subir(byte[] contenido, String contentType, String nombreArchivo,
                         Long productoId, Long usuarioId);
    List<DisenoResponse> listarPorUsuario(Long usuarioId);
    void eliminar(Long id, Long usuarioId);
    int eliminarTodos(Long usuarioId);

    // ---- Publicación pública (galería de diseños de usuarios) ----

    /** Envía un diseño propio a moderación para que sea visible en la galería. */
    DisenoResponse publicar(Long disenoId, String titulo, Long usuarioId);

    /** Lista los diseños públicos aprobados, ordenables por popularidad. */
    Page<DisenoPublicoResponse> listarPublicos(String orden, Pageable pageable, Long usuarioId);

    /** Obtiene un diseño público por id (con métricas). */
    DisenoPublicoResponse obtenerPublico(Long id, Long usuarioId);

    /** Da "me gusta" a un diseño público. Devuelve true si se registró. */
    boolean darMeGusta(Long disenoId, Long usuarioId);

    /** Quita el "me gusta". Devuelve true si existía. */
    boolean quitarMeGusta(Long disenoId, Long usuarioId);

    /** ¿El usuario ya dio "me gusta" a este diseño? */
    boolean estaEnMeGusta(Long disenoId, Long usuarioId);
}

File: frontend/src/services/disenoService.js
import api from "./api";

export async function generarDiseno({ prompt, productoId, estilo, imagenReferencia }) {
  const { data } = await api.post("/disenos/generar", {
    prompt,
    productoId,
    estilo,
    imagenReferencia,
  });
  return data;
}

export async function listarDisenosPorUsuario() {
  const { data } = await api.get("/disenos/usuario");
  return data;
}

/**
 * Sube el PNG final compuesto en el editor manual (Konva) al backend.
 * El backend valida que sea imagen, la guarda en Supabase Storage y crea un
 * registro Diseno con origen=USUARIO. El token JWT lo inyecta el interceptor
 * de `api` automáticamente (Authorization: Bearer <token>).
 */
export async function subirDiseno({ file, productoId }) {
  const formData = new FormData();
  formData.append("file", file);
  formData.append("productoId", productoId);
  formData.append("origen", "USUARIO");

  const { data } = await api.post("/disenos/subir", formData);
  return data;
}

/**
 * Agrega un diseño ya creado al carrito (POST /api/detalle-carrito).
 * El backend verifica que el diseño pertenezca al usuario autenticado.
 */
export async function agregarDisenoAlCarrito({
  carritoId,
  productoId,
  disenoId,
  cantidad = 1,
}) {
  const { data } = await api.post("/detalle-carrito", {
    carritoId,
    productoId,
    cantidad,
    disenoId,
  });
  return data;
}

export async function eliminarDiseno(id) {
  await api.delete(`/disenos/${id}`);
}

export async function eliminarTodosDisenos() {
  const { data } = await api.delete("/disenos");
  return data;
}

// ---------------------------------------------------------------------------
// Publicación pública (galería de diseños de usuarios)
// ---------------------------------------------------------------------------

/**
 * Envía un diseño propio a la galería pública. El admin debe aprobarlo
 * antes de que sea visible para todos.
 */
export async function publicarDiseno({ id, titulo }) {
  const { data } = await api.post(`/disenos/${id}/publicar`, { titulo });
  return data;
}

/**
 * Lista los diseños públicos aprobados (galería).
 * orden: "recientes" (default) | "megusta" | "usos"
 */
export async function listarDisenosPublicos({ page = 0, size = 12, orden = "recientes" } = {}) {
  const { data } = await api.get("/disenos/publicos", {
    params: { page, size, orden },
  });
  return data;
}

export async function obtenerDisenoPublico(id) {
  const { data } = await api.get(`/disenos/publicos/${id}`);
  return data;
}

/** Da "me gusta" a un diseño público (requiere sesión). */
export async function darMeGustaDiseno(id) {
  const { data } = await api.post(`/disenos/${id}/megusta`);
  return data;
}

/** Quita el "me gusta" (requiere sesión). */
export async function quitarMeGustaDiseno(id) {
  const { data } = await api.delete(`/disenos/${id}/megusta`);
  return data;
}

/** Estado del "me gusta" del usuario autenticado sobre un diseño. */
export async function estadoMeGustaDiseno(id) {
  const { data } = await api.get(`/disenos/${id}/megusta`);
  return data;
}
